### ISSUE TITLE
Time for training

### ISSUE THREAD

**Issue by DbrRoxane** [shmsw25/qa-hard-em#10]
Hi,

I have downloaded the preprocessed data from your repo and tried to run the training on TriviaQA with the hard-em loss with `./run.sh triviaqa hard-em 4000`
I had to change the BS for training to 32 (instead of 48) because I had some error occurring. I have two GPUs, I checked that the code effectively use it but still, after 2 days the epoch 0 was still not over... (I was on the 8,000 step). I guess it is not nomal? :o 
So here are my questions : is it normal? if no, do you have a clue what is going wrong? 
I need to train the model because I want to use the best ckpt to test on other data. So if it is normal, could you release the best checkpoint?

Regards,

**Comment by shmsw25** [comment]
Hi @DbrRoxane, that sounds weird. The entire training took round 2 days for me and I typically reached 40k-80k steps. The fact that you were on the 8k steps after 2 days seems weird, even with the smaller batch sizes... Did you check the model is properly on the gpus?

**Comment by DbrRoxane** [comment]
Hi @shmsw25 . Yes I reran the training and it is the same as before. The model use the both GPUs to train, I double checked.
What looks actually weird is that the most time consuming activity seems to be the prediction and not the training.  Each eval_period, the prediction takes something like two hours, as you can see below
![image](https://user-images.githubusercontent.com/32749371/79193285-9c0c4f00-7e2a-11ea-9f12-b0c085f96909.png)

I changed the number of step (eval_period) from 1k to 100k for the moment. Is it going to impact the results in any way ? 
Also, do you have the best model checkpoints that we can download?

**Comment by shmsw25** [comment]
Thanks for getting back. I am not sure why it takes so long, but I will investigate. It won'y impact the result so much, although less frequently examining the checkpoint might lead to miss the best checkpoint.

I don't have the best checkpoint right now, but I can run the code again to get it (can't promise when, but maybe in 2 weeks). Do you need TriviaQA model?

**Comment by DbrRoxane** [comment]
Hi, thank you so much, yes it is for TriviaQA model. 
I also continue to investigate and make some experiments, I'll keep you informed if I find smth interesting.

**Comment by Liuhongyu12138** [comment]
hi~ I have the same problem that the eval time during training is too long. In my gpu it takes about 1.5 hours for each eval_period. Have you solve this problems or just increase the eval_time?

**Comment by DbrRoxane** [comment]
Hi @Liuhongyu12138 , I partially solved it by changing the eval_period to 100k (in main.py , parser.add_argument('--eval_period', type=int, default=10000) or manually in the run.sh)
Because, yes, indeed, this is predicting on devset and saving the model which is the longest, not the training by itself.

**Comment by Liuhongyu12138** [comment]
@DbrRoxane Thanks! Actually I have changed eval_period to a larger number like 50k. And at the same time I changed the wait_step to a smaller number like 3 rather than the original 12. Because 12 means that you need at least 12 * eval time (like 2 hours for every eval, total 24 hours) to break the training for best checkpoint. That's too long for me, but I'm afraid that I will missing the best checkpoint when using the smaller wait_step.
Have you reproduce the similar EM result in triviaqa (openqa setting) ? And can you tell me your hyperparameters and the best checkpoint's training steps ? Thanks a lot !

**Comment by DbrRoxane** [comment]
Hi,

So I could finally use a machine with 8GPUs so I ran the code with eval_period = 5000;
But the other hyperparam are the same ,
Oh and I ran first-only code, not the other loss since it fits to my data
The best model I have is recorded after the 85k step, here is my output before saving
Step 85000 Train loss 2.11 EM 44.26 F1 49.09 on epoch=7

I tested the model with test data, executing the last command of run.sh , which gives me actually better results than in the paper :) 
n=10, EM=47.66
n=20, EM=49.21
n=40, EM=49.60
n=80, EM=49.08

I can send the checkpoint if you like!

**Comment by Liuhongyu12138** [comment]
@DbrRoxane Wow ! That's very useful information for me! Thank you !!! Do I need to modify the source code if I want to run this code in multi-gpu ? or just add CUDA_VISIBLE_DEVICES=0, 1, 2, 3 in the run.sh for four gpu training and need't change any code in the main.py ?

**Comment by Liuhongyu12138** [comment]
@DbrRoxane Wow ! That's very useful information for me! Thank you !!! 
I have't run this code in multi-gpu. Do I need to modify the source code(main.py) if I want to run this code in multi-gpu ? or just add CUDA_VISIBLE_DEVICES=0, 1, 2, 3 in the run.sh for four gpu training and need't change any code in the main.py ?

**Comment by DbrRoxane** [comment]
You're very welcome! It detects it automatically I don't think I had to change anything else, you can ensure by displaying watch nvidia-smi or so on your shell.

**Comment by Liuhongyu12138** [comment]
Thanks! That's a very useful information for me : ) Have a nice day !

**Comment by sukannyapurkayastha** [comment]
Hi, thanks for making your code public. I was wondering if it is possible to get the documentation on how you implement the mml and hard_em model? Actually I am not being able to understand some parts of your logic in the code?

**Comment by DbrRoxane** [comment]
Hi @sukannyapurkayastha  I think you should ask your question in a new issue or sending email since it is not related to the train for training.
